### PR TITLE
Prepare for implementation in `aquadoggo`

### PR DESCRIPTION
**Pull request by sandreae**
These changes make implementation of the new validation features and API changes more pleasant / logical in `aquadoggo`. Some patterns (particularly around the storage types) needed to change, all for the better, and I think in general there is a nicer flow now. The main point which has changed is that all data now remains as core `p2panda-rs` types (`Entry`, `EncodedEntey`, `VerifiedOperation`) right up until it is in the store. We then use the storage types for data coming out of the store and being handled. These types now require the same traits implemented on the core types, so can be handled in much the same way. 

- rename `AsStorageEntry` trait to `EntryWithOperation` 
- `EntryWithOperation` has access to an entry, it's encoded form, and optionally it's payload (i scrapped the idea of removing the payload here, added more complex changes that are noooooot needed here).
-  introduced `AsEntry` and `AsEncodedEntry` traits which as well as being implemented on their respective structs and are required traits on `EntryWithOperation`
- `insert_entry` and `insert_operation` on `EntryStore` and `OperationStore` now _do not_ accept app implemented storage structs, but rathe p2panda-rs types. This removes the need for a constructor on the storage types and means we can rely on our own _hard core_ validated types without adding a layer of conversion.
- entries and operations which come out of the store _do_ implement storage traits and are app defined structs.
- conversion traits from storage types into p2panda types
- `DocumentBuilder` is generic over anything that implements `AsVerifiedOperation`

## 📋 Checklist

- [ ] Add tests that cover your changes
- [ ] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [ ] Link this PR to any issues it closes
- [ ] New files contain a SPDX license header
- [ ] Check if descriptions and terminology match `handbook` content (and visa-versa)

### CODE REVIEW

File: p2panda-rs/src/document/document_id.rs [hunk 1]
```diff
@@ -23,7 +25,7 @@ use crate::{Human, Validate};
 ///                         \
 ///                          \__ [UPDATE] (Hash: "eff..")
 /// ```
-#[derive(Clone, Debug, StdHash, Ord, PartialOrd, Eq, PartialEq)]
+#[derive(Clone, Debug, StdHash, Ord, PartialOrd, Eq, PartialEq, Serialize, Deserialize)]
```
Comment (adzialocha) [inline p2panda-rs/src/document/document_id.rs:28]: We can't derive `Deserialize` as this introduces unchecked values. You can implement `Deserialize` yourself and validate the String, see `Hash` or `DocumentViewId` for examples :+1:

Comment (sandreae) [reply p2panda-rs/src/document/document_id.rs:28]: ah, ok... Thanks!

Comment (adzialocha) [reply p2panda-rs/src/document/document_id.rs:28]: And maybe you missed my comment in the other PR but maybe double check if we need serde for scalars ✌️

File: p2panda-rs/src/operation/verified_operation.rs [hunk 2]
```diff
@@ -2,9 +2,14 @@
...
 use std::hash::Hash as StdHash;
 
+use crate::document::DocumentViewId;
 use crate::identity::Author;
 use crate::operation::traits::AsVerifiedOperation;
 use crate::operation::{Operation, OperationId};
+use crate::schema::SchemaId;
+
+use super::traits::AsOperation;
+use super::{OperationAction, OperationFields, OperationVersion};
```
Comment (adzialocha) [inline p2panda-rs/src/operation/verified_operation.rs:12]: Not use super here

File: p2panda-rs/src/storage_provider/traits/storage_provider.rs [hunk 3]
```diff
@@ -3,32 +3,33 @@
...
+use crate::entry::traits::{AsEncodedEntry, AsEntry};
 use crate::hash::Hash;
 use crate::operation::traits::AsVerifiedOperation;
-use crate::storage_provider::traits::{
-    AsStorageEntry, AsStorageLog, EntryStore, LogStore, OperationStore,
-};
+use crate::storage_provider::traits::{AsStorageLog, EntryStore, LogStore, OperationStore};
 use crate::storage_provider::utils::Result;
 
+use super::EntryWithOperation;
```
Comment (adzialocha) [inline p2panda-rs/src/storage_provider/traits/storage_provider.rs:12]: No super here 👍

**Review by adzialocha — APPROVED** [review]
Cool, I think this looks all good! Maybe you remove the super-level imports first, but otherwise it's ready.

File: p2panda-rs/src/storage_provider/traits/storage_provider.rs (hunk 3 above)
Comment (sandreae) [reply p2panda-rs/src/storage_provider/traits/storage_provider.rs:12]: super duper!

Comment (adzialocha) [reply p2panda-rs/src/storage_provider/traits/storage_provider.rs:12]: Nothing is super!

Comment (adzialocha) [reply p2panda-rs/src/storage_provider/traits/storage_provider.rs:12]: > super duper!

Super ✌️

**Comment by sandreae** [comment]
I'll just fix the js tests before merging :+1:

**Pull request merged**